runtime/runtime-core/common/test/aws/smithy/kotlin/runtime/util/ReadThroughCacheTest.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.util
+
+import aws.smithy.kotlin.runtime.time.ManualClock
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.runTest
+import org.junit.jupiter.api.Assertions.assertEquals
+import kotlin.test.Test
+import kotlin.time.Duration.Companion.minutes
+import kotlin.time.Duration.Companion.seconds
+
+@OptIn(ExperimentalCoroutinesApi::class)
+class ReadThroughCacheTest {
+    @Test
+    fun testReadThrough() = runTest {
+        val clock = ManualClock()
+        var counter = 0
+        fun uncachedValue() = ExpiringValue(counter++, clock.now() + 2.seconds)
+        val cache = ReadThroughCache<String, Int>(1.minutes, clock)
+
+        // Basic read through
+        assertEquals(0, cache.get("a") { uncachedValue() })
+        assertEquals(1, cache.get("b") { uncachedValue() })
+
+        // Basic cache verification
+        assertEquals(0, cache.get("a") { uncachedValue() })
+        assertEquals(1, cache.get("b") { uncachedValue() })
```

Comment:
here instead of `uncachedValue()` it could be better to pass a function that throws an exception, to check that the cached value is read instead of the pass-through function being invoked